### ISSUE TITLE
Leading Zeros are not displayed with TEXT format

### ISSUE THREAD

**Issue by NicoHood** [jmcnamara/excel-writer-xlsx#146]
Hi, I try to add numbers with leading zeros to my table. Example: "0100".
However excel crosses out the leading zeros. To avoid this, you can use the [TEXT format which prevents this](http://stackoverflow.com/questions/2261787/how-to-prevent-automatic-truncation-of-leading-zeros-in-excel-cell).

The problem is, that with this perl script the leading zeros are still deleted, even with TEXT format. I cannot use any other format (like "0000")  because I never know if the data is text or a number, so I just want to paste it as Text "as it is".

This happens with the normal write() function and with the add_table() function.

```perl
#!/usr/bin/perl -w

use strict;
use Excel::Writer::XLSX;

my $workbook = Excel::Writer::XLSX->new( 'formats.xlsx' );
my $worksheet = $workbook->add_worksheet();

my $format = $workbook->add_format();
my $numFormat = 0x31; # @ (TEXT)
$format->set_num_format( $numFormat );

$worksheet->write( 0, 0, '0100', $format );

__END__
```

With VBA you can do this real easy by using:
```vba
Sub Macro1()
    Selection.NumberFormat = "@"
    ActiveCell.FormulaR1C1 = "0100"
End Sub

' Even works with tables
Sub Macro2()
        Dim RNG As Range
        Set RNG = Range(Cells(1, 1), Cells(2, 1)).Offset(0, 0)
        
        RNG.NumberFormat = "@"
        RNG.Value = "0111"

        ' Format table
        Dim tbl As ListObject
        Set tbl = ActiveSheet.ListObjects.Add(xlSrcRange, RNG, , xlYes)
        tbl.TableStyle = "TableStyleMedium3"
End Sub
```

Is this intended/a limit of what you can do with perl or a bug?
Any Idea how to fix this?

**Comment by jmcnamara** [comment]
The main issue is that the `write()` method converts the string to a number and thus the leading zero is lost. There are several workarounds but the simplest is probably to use the `keep_leading_zeros()` function.

I'll let the [docs](http://search.cpan.org/dist/Excel-Writer-XLSX/lib/Excel/Writer/XLSX.pm#keep_leading_zeros%28%29) explain this one: :-)


#### `keep_leading_zeros()`

>This method changes the default handling of integers with leading zeros when using the write() method.

>The write() method uses regular expressions to determine what type of data to write to an Excel worksheet. If the data looks like a number it writes a number using `write_number()`. One problem with this approach is that occasionally data looks like a number but you don't want it treated as a number.

>Zip codes and ID numbers, for example, often start with a leading zero. If you write this data as a number then the leading zero(s) will be stripped. This is the also the default behaviour when you enter data manually in Excel.

>To get around this you can use one of three options. Write a formatted number, write the number as a string or use the `keep_leading_zeros()` method to change the default behaviour of `write()`:

```perl
# Implicitly write a number, the leading zero is removed: 1209
$worksheet->write( 'A1', '01209' );

# Write a zero padded number using a format: 01209
my $format1 = $workbook->add_format( num_format => '00000' );
$worksheet->write( 'A2', '01209', $format1 );

# Write explicitly as a string: 01209
$worksheet->write_string( 'A3', '01209' );

# Write implicitly as a string: 01209
$worksheet->keep_leading_zeros();
$worksheet->write( 'A4', '01209' );
```
>The above code would generate a worksheet that looked like the following:

     -----------------------------------------------------------
    |   |     A     |     B     |     C     |     D     | ...
     -----------------------------------------------------------
    | 1 |      1209 |           |           |           | ...
    | 2 |     01209 |           |           |           | ...
    | 3 | 01209     |           |           |           | ...
    | 4 | 01209     |           |           |   

>The examples are on different sides of the cells due to the fact that Excel displays strings with a left justification and numbers with a right justification by default. You can change this by using a format to justify the data,

**Comment by NicoHood** [comment]
Thanks, I totally missed that.

I need this inside a table `add_table()`. So I guess the only option I can use here is the `$worksheet->keep_leading_zeros();`? I tried that and it works great so far. Since there is no write_string() or so this is the only option?

Another question real quick:
I am reading a csv file and place it inside the .xlsx file. Then I want to format it as table. First I tried to write() the data and place a table on top of that. But the header is missing then. So I wrote the data + header in an array and it works fine now. Is this the only option I have, cant I just use the existing data as header?

Is there a way to center the header data as well? (I applied the center style to all colums but the header was kept left aligned).

Thx :)

**Comment by jmcnamara** [comment]
> I tried that and it works great so far. Since there is no write_string() or so this is the only option?

If it works then stick with it, there shouldn't be any unwanted side-effects. If you prefer to explicitly set the type of the data in the table cells you can write the data separately, like example 3 in the [tables.pl](https://github.com/jmcnamara/excel-writer-xlsx/blob/master/examples/tables.pl). 

> cant I just use the existing data as header?

Yes. If you use the table column option (see the [docs](http://search.cpan.org/dist/Excel-Writer-XLSX/lib/Excel/Writer/XLSX.pm#TABLES_IN_EXCEL)):

```perl
    $worksheet->add_table(
        'B3:F7',
        {
            data    => $data,
            columns => [
                { header => 'Product' },
                { header => 'Quarter 1' },
                { header => 'Quarter 2' },
                { header => 'Quarter 3' },
                { header => 'Quarter 4' },
            ]
        }
    );
```

> Is there a way to center the header data as well? (I applied the center style to all colums but the header was kept left aligned).

This is becoming a FAQ and it probably means that I need some way to format the header. In the meantime you can work around it using the technique shown here: https://github.com/jmcnamara/XlsxWriter/issues/287

**Comment by NicoHood** [comment]
> cant I just use the existing data as header?

Yes I finally did it like that, because there is no other way to do it.
I'll try to explain better: Cant i set a table over cell data that has been written with write()?
I first tried to write() the data and put a table on top of that. That worked for the data but the header got overwritten by "column1" etc. Like the example you showed, but with the header. I did it in the order write->table, but the example did it the other way around. Maybe that was my fault/a bug.

**Comment by jmcnamara** [comment]
> I'll try to explain better: Cant i set a table over cell data that has been written with write()?

Yes for the data. No for the header.

The header data must always be added via `columns` as shown above.


> I did it in the order write->table, but the example did it the other way around. Maybe that was my fault/a bug.

I order shouldn't matter. But if you write the header data separately you must also write it via `columns` as shown in the Python example I linked to above.

There are some gotchas in that approach so read my comments in that thread carefully.

John

**Comment by jmcnamara** [comment]
The option to add formatting to table headers has been added. See #148.

**Comment by jmcnamara** [comment]
Closed in version 0.86 on CPAN.

Thanks.

**Comment by eslint-config-airbnb** [comment]
Ponder this solution



https://stackdev.space/questions/combining-a-range-to-create-a-nicely-formatted-string-with-the-help-of-arrays
